File: android/activitysources/src/main/java/com/fjuul/sdk/activitysources/workers/HCIntradaySyncWorker.java
package com.fjuul.sdk.activitysources.workers;

import java.util.concurrent.ExecutionException;

import com.fjuul.sdk.activitysources.entities.ActivitySourceConnection;
import com.fjuul.sdk.activitysources.entities.ActivitySourcesManager;
import com.fjuul.sdk.activitysources.entities.FitnessMetricsType;
import com.fjuul.sdk.activitysources.entities.HealthConnectActivitySource;
import com.fjuul.sdk.activitysources.entities.HealthConnectIntradaySyncOptions;
import com.google.android.gms.tasks.TaskCompletionSource;
import com.google.android.gms.tasks.Tasks;

import android.annotation.SuppressLint;
import android.content.Context;
import androidx.annotation.NonNull;
import androidx.work.WorkerParameters;

public class HCIntradaySyncWorker extends HCSyncWorker {
    public static final String KEY_INTRADAY_METRICS_ARG = "INTRADAY_METRICS";

    public HCIntradaySyncWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {
        super(context, workerParams);
    }

    @SuppressLint("NewApi")
    @NonNull
    @Override
    public Result doWork() {
        final ActivitySourcesManager sourcesManager = getOrInitializeActivitySourcesManager();
        final ActivitySourceConnection hcConnection = getGoogleHealthConnectActivitySourceConnection(sourcesManager);
        if (hcConnection == null) {
            // TODO: cancel next scheduled tasks because there is not current hc connection
            // NOTE: currently, the task will be canceled on the next initialization of ActivitySourcesManager
            return Result.success();
        }
        final HealthConnectActivitySource hcSource =
            ((HealthConnectActivitySource) hcConnection.getActivitySource());
        final TaskCompletionSource<Void> taskCompletionSource = new TaskCompletionSource<>();
        final HealthConnectIntradaySyncOptions syncOptions = buildIntradaySyncOptions();
        hcSource.syncIntradayMetrics(syncOptions, (result -> {
            if (result.isError() && result.getError() instanceof Exception) {
                taskCompletionSource.trySetException((Exception) result.getError());
                return;
            }
            taskCompletionSource.trySetResult(null);
        }));
        try {
            Tasks.await(taskCompletionSource.getTask());
            return Result.success();
        } catch (ExecutionException exception) {
            Exception originCause = (Exception) exception.getCause();
            // TODO: figure out what caused the exception and handle this
        } catch (InterruptedException e) {}
        return Result.failure();
    }

    @SuppressLint("NewApi")
    private HealthConnectIntradaySyncOptions buildIntradaySyncOptions() {
        final String[] rawIntradayMetrics = getInputData().getStringArray(KEY_INTRADAY_METRICS_ARG);
        final HealthConnectIntradaySyncOptions.Builder syncOptionsBuilder =
            new HealthConnectIntradaySyncOptions.Builder();
        for (final String rawIntradayMetric : rawIntradayMetrics) {
            try {
                final FitnessMetricsType metric = FitnessMetricsType.valueOf(rawIntradayMetric);
                if (metric != null) {
                    syncOptionsBuilder.include(metric);
                }
            } catch (Exception e) {}
        }
        return syncOptionsBuilder.build();
    }
}

File: android/activitysources/src/main/java/com/fjuul/sdk/activitysources/workers/HCIntradaySyncWorker.kt
package com.fjuul.sdk.activitysources.workers

import android.content.Context
import androidx.work.WorkerParameters
import com.fjuul.sdk.activitysources.entities.FitnessMetricsType
import com.fjuul.sdk.activitysources.entities.HealthConnectActivitySource
import com.fjuul.sdk.activitysources.entities.internal.healthconnect.HealthConnectSyncOptions
import com.fjuul.sdk.core.utils.Logger
import com.google.android.gms.tasks.TaskCompletionSource
import com.google.android.gms.tasks.Tasks

class HCIntradaySyncWorker(context: Context, workerParams: WorkerParameters) :
    HCSyncWorker(context, workerParams) {

    override suspend fun doWork(): Result {
        val hcConnection = getHealthConnectActivitySourceConnection(activitySourcesManager)
        if (hcConnection == null) {
            return Result.success()
        }
        val hcSource = (hcConnection.activitySource as HealthConnectActivitySource)
        val permissionManager = hcSource.getPermissionManager()

        if (permissionManager.checkBackgroundPermission()) {
            val taskCompletionSource = TaskCompletionSource<Void?>()
            val syncOptions = buildIntradaySyncOptions()
            hcSource.syncIntraday(syncOptions) { result ->
                if (result.isError && result.error is Exception) {
                    taskCompletionSource.trySetException(result.error as Exception)
                    return@syncIntraday
                }
                taskCompletionSource.trySetResult(null)
            }

            try {
                Tasks.await<Void?>(taskCompletionSource.getTask())
                return Result.success()
            } catch (e: Exception) {
                Logger.get().e(e, "Exception during intraday sync")
            }
        }
        return Result.failure()
    }

    private fun buildIntradaySyncOptions(): HealthConnectSyncOptions {
        val metrics = mutableSetOf<FitnessMetricsType>()
        val rawIntradayMetrics = inputData.getStringArray(KEY_HC_INTRADAY_METRICS) ?: emptyArray()
        for (rawIntradayMetric in rawIntradayMetrics) {
            try {
                val metric = FitnessMetricsType.valueOf(rawIntradayMetric)
                metrics.add(metric)
            } catch (_: Exception) {
            }
        }
        return HealthConnectSyncOptions(metrics)
    }

    companion object {
        const val KEY_HC_INTRADAY_METRICS: String = "HC_INTRADAY_METRICS"
    }
}
